packages/api/endpoints/granules.js

```diff
@@ -200,6 +206,147 @@ const create = async (req, res) => {
   });
 };
 
+/**
+ * Update existing granule *or* create new granule
+ *
+ * @param {Object} req - express request object
+ * @param {Object} res - express response object
+ * @returns {Promise<Object>} promise of an express response object.
+ */
+const patchGranule = async (req, res) => {
+  const {
+    granulePgModel = new GranulePgModel(),
+    collectionPgModel = new CollectionPgModel(),
+    knex = await getKnexClient(),
+    esClient = await Search.es(),
+    updateGranuleFromApiMethod = updateGranuleFromApi,
+  } = req.testContext || {};
+  let apiGranule = req.body || {};
+  let pgCollection;
+
+  if (!apiGranule.collectionId) {
+    res.boom.badRequest('Granule update must include a valid CollectionId');
+  }
+
+  try {
+    pgCollection = await collectionPgModel.get(
+      knex,
+      deconstructCollectionId(apiGranule.collectionId)
+    );
+  } catch (error) {
+    if (error instanceof RecordDoesNotExist) {
+      log.error(
+        `granule collectionId ${apiGranule.collectionId} does not exist, cannot update granule`
+      );
+      res.boom.badRequest(
+        `granule collectionId ${apiGranule.collectionId} invalid`
+      );
+    } else {
+      throw error;
+    }
+  }
+
+  // TODO: CUMULUS-3017 - Remove this unique collectionId condition
+  // Check if granuleId exists across another collection
+  const granulesByGranuleId = await getGranulesByGranuleId(
+    knex,
+    apiGranule.granuleId
+  );
+  const granuleExistsAcrossCollection = granulesByGranuleId.some(
+    (g) => g.collection_cumulus_id !== pgCollection.cumulus_id
+  );
+  if (granuleExistsAcrossCollection) {
+    log.error(
+      'Could not update or write granule, collectionId is not modifiable.'
+    );
+    return res.boom.conflict(
+      `Modifying collectionId for a granule is not allowed. Write for granuleId: ${apiGranule.granuleId} failed.`
+    );
+  }
+
+  let isNewRecord = false;
+  try {
+    await granulePgModel.get(knex, {
+      granule_id: apiGranule.granuleId,
+      collection_cumulus_id: pgCollection.cumulus_id,
+    }); // TODO this should do a select count, not a full record get
```

Comment:
TODO?

Reply:
Resolved - out of scope/separate ticket/still need.